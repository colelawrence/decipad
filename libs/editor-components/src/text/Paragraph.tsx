import {
  COLUMN_KINDS,
  MyEditor,
  MyElement,
  PlateComponent,
  useTEditorRef,
} from '@decipad/editor-types';
import {
  getRangeSafe,
  isDragAndDropHorizontal,
  useNodePath,
} from '@decipad/editor-utils';
import {
  useEditorChangeState,
  useIsEditorReadOnly,
} from '@decipad/react-contexts';
import { ParagraphPlaceholder, Paragraph as UIParagraph } from '@decipad/ui';
import { isElementEmpty, isSelectionExpanded } from '@udecode/plate';
import { Range } from 'slate';
import { ReactEditor } from 'slate-react';
import { DraggableBlock } from '../block-management';
import { useDragAndDropGetAxis, useDragAndDropOnDrop } from '../hooks';
import { useTurnIntoProps } from '../utils';

const isSelected = (editor: MyEditor, element: MyElement) => {
  if (!editor.selection) {
    return false;
  }
  const path = ReactEditor.findPath(editor as ReactEditor, element);
  if (!path) {
    return false;
  }
  const range = getRangeSafe(editor, path);
  return !!(range && Range.intersection(range, editor.selection));
};

export const Paragraph: PlateComponent = ({
  attributes,
  children,
  element,
}) => {
  if (!element) {
    throw new Error('Paragraph is not a leaf');
  }
  const readOnly = useIsEditorReadOnly();

  // Performance improvement as opposed to use something like `useTEditorState()`.
  // We couldn't use `useTEditorRef()` because we need the paragraph to re-render
  // when the amount of editor children change.
  const showPlaceHolder = useEditorChangeState(
    (editor) =>
      (editor.children.length === 2 && isElementEmpty(editor, element)) ||
      (isElementEmpty(editor, element) &&
        isSelected(editor, element) &&
        !isSelectionExpanded(editor)),
    false,
    {
      debounceTimeMs: 0,
    }
  );

  const placeholder =
    readOnly || !showPlaceHolder ? undefined : <ParagraphPlaceholder />;

  const turnIntoProps = useTurnIntoProps(element);
  const editor = useTEditorRef();
  const path = useNodePath(element);
  const isHorizontal = isDragAndDropHorizontal(false, editor, path);
  const getAxis = useDragAndDropGetAxis({ isHorizontal });
  const onDrop = useDragAndDropOnDrop({ editor, element, path, isHorizontal });

  return (
    <DraggableBlock
      blockKind="paragraph"
      element={element}
      accept={isHorizontal ? COLUMN_KINDS : undefined}
      getAxis={getAxis}
      onDrop={onDrop}
      {...turnIntoProps}
      {...attributes}
    >
      <UIParagraph placeholder={placeholder}>{children}</UIParagraph>
    </DraggableBlock>
  );
};
