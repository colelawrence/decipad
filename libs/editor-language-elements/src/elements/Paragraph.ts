import { ELEMENT_PARAGRAPH, MARK_MAGICNUMBER } from '@decipad/editor-types';
import { assertElementType, magicNumberId } from '@decipad/editor-utils';
import { getNodeString } from '@udecode/plate';
import { weakMapMemoizeInteractiveElementOutput } from '../utils/weakMapMemoizeInteractiveElementOutput';
import { InteractiveLanguageElement } from '../types';

export const Paragraph: InteractiveLanguageElement = {
  type: ELEMENT_PARAGRAPH,
  isStructural: true,
  getUnparsedBlockFromElement: weakMapMemoizeInteractiveElementOutput(
    (_editor, element) => {
      assertElementType(element, ELEMENT_PARAGRAPH);

      return element.children.flatMap((child, index) => {
        if (MARK_MAGICNUMBER in child) {
          return {
            type: 'unparsed-block',
            id: magicNumberId(element, index),
            source: getNodeString(child),
          };
        }
        return [];
      });
    }
  ),
};
