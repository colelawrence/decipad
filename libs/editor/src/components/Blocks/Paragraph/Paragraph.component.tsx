import { css } from '@emotion/react';
import { SlatePluginComponent } from '@udecode/slate-plugins';

const pStyles = css({
  lineHeight: '1.75',
  color: '#3E3E42',
  padding: '8px 0',
  position: 'relative',
  '& ::selection': {
    backgroundColor: 'rgba(196, 202, 251, 0.5)',
  },
});

export const Paragraph: SlatePluginComponent = ({
  attributes,
  children,
  nodeProps,
}) => {
  const root = nodeProps!.styles.root;
  return (
    <p css={[pStyles, root.css]} {...attributes} {...nodeProps}>
      {children}
    </p>
  );
};
