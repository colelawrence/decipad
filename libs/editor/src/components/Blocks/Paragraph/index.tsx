import styled from '@emotion/styled';
import { SlatePluginComponent } from '@udecode/slate-plugins';
import React from 'react';

const Element = styled('p')({
  lineHeight: '1.75',
  color: '#3E3E42',
  padding: '8px 0',
  position: 'relative',
  '& ::selection': {
    backgroundColor: 'rgba(196, 202, 251, 0.5)',
  },
});

export const Paragraph: SlatePluginComponent = ({ attributes, children }) => {
  return <Element {...attributes}>{children}</Element>;
};
