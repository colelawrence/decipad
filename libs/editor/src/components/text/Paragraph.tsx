import { atoms } from '@decipad/ui';
import { useEditorState, isSelectionExpanded } from '@udecode/plate';
import { Editor } from 'slate';
import { useSelected } from 'slate-react';
import { PlateComponent } from '../../types';

export const Paragraph: PlateComponent = ({
  attributes,
  children,
  element,
}) => {
  const editor = useEditorState();
  const selected = useSelected();

  if (!element) {
    throw new Error('Paragraph is not a leaf');
  }

  return (
    <div {...attributes}>
      <atoms.Paragraph
        placeholder={
          Editor.isEmpty(editor, element) &&
          selected &&
          !isSelectionExpanded(editor)
            ? 'Type “/” for commands or write text'
            : undefined
        }
      >
        {children}
      </atoms.Paragraph>
    </div>
  );
};
