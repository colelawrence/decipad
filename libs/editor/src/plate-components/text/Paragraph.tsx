import { types } from '@decipad/editor-config';
import { atoms } from '@decipad/ui';
import { isSelectionExpanded, useEditorState } from '@udecode/plate';
import { Editor } from 'slate';
import { useReadOnly, useSelected } from 'slate-react';
import { DraggableBlock } from '../../components';

export const Paragraph: types.PlateComponent = ({
  attributes,
  children,
  element,
}) => {
  const editor = useEditorState();
  const selected = useSelected();
  const readOnly = useReadOnly();

  if (!element) {
    throw new Error('Paragraph is not a leaf');
  }

  return (
    <div {...attributes}>
      <DraggableBlock blockKind="paragraph" element={element}>
        <atoms.Paragraph
          placeholder={
            Editor.isEmpty(editor, element) &&
            selected &&
            !isSelectionExpanded(editor) &&
            !readOnly
              ? 'Type “/” for commands or write text'
              : undefined
          }
        >
          {children}
        </atoms.Paragraph>
      </DraggableBlock>
    </div>
  );
};
