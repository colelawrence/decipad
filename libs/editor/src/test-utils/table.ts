import {
  ELEMENT_TABLE,
  ELEMENT_TH,
  ELEMENT_TR,
  ELEMENT_TD,
} from '@udecode/plate';
import {
  InteractiveTable,
  TableCellType,
  Td,
  Th,
} from '../plugins/InteractiveTable/table';
import {
  ELEMENT_TABLE_CAPTION,
  ELEMENT_THEAD,
  ELEMENT_HEAD_TR,
  ELEMENT_TBODY,
} from '../utils/elementTypes';

/**
 * Creates a row with 1 named column and {@param numberOfRows} rows of non-empty cells.
 */
export const createTable = (numberOfRows: number): InteractiveTable => ({
  type: ELEMENT_TABLE,
  children: [
    {
      type: ELEMENT_TABLE_CAPTION,
      children: [{ text: 'Table' }],
    },
    {
      type: ELEMENT_THEAD,
      children: [
        {
          type: ELEMENT_HEAD_TR,
          children: [
            {
              type: ELEMENT_TH,
              attributes: { cellType: 'string' },
              children: [{ text: 'Col 1' }],
            },
          ],
        },
      ],
    },
    {
      type: ELEMENT_TBODY,
      children: Array.from({ length: numberOfRows }, (_, i) => ({
        type: ELEMENT_TR,
        children: [
          {
            type: ELEMENT_TD,
            attributes: { cellType: 'string' },
            children: [{ text: `Col 1 Row ${i + 1}` }],
          },
        ],
      })),
    },
  ],
});

export const createTableWithTypeAndData = (
  cellType: TableCellType,
  data: string[]
): InteractiveTable => ({
  type: ELEMENT_TABLE,
  children: [
    {
      type: ELEMENT_TABLE_CAPTION,
      children: [{ text: 'Table' }],
    },
    {
      type: ELEMENT_THEAD,
      children: [
        {
          type: ELEMENT_HEAD_TR,
          children: [
            {
              type: ELEMENT_TH,
              attributes: { cellType },
              children: [{ text: 'Col 1' }],
            },
          ],
        },
      ],
    },
    {
      type: ELEMENT_TBODY,
      children: data.map((text) => ({
        type: ELEMENT_TR,
        children: [
          {
            type: ELEMENT_TD,
            attributes: { cellType },
            children: [{ text }],
          },
        ],
      })),
    },
  ],
});

export const printTable = ({
  children: [_, thead, tbody],
}: InteractiveTable): string[] => {
  const tdsToStrings = (tds: (Th | Td)[]) =>
    tds.map((td) => td.children[0].text).join(' | ');

  return [
    tdsToStrings(thead.children[0].children),
    ...tbody.children.map((tr) => tdsToStrings(tr.children)),
  ];
};
