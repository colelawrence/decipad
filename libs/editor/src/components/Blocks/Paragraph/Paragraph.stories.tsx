import React from 'react';
import { Paragraph } from './Paragraph.component';

export default {
  title: 'Editor/Blocks/Paragraph',
  component: Paragraph,
};

export const Default = (args: { text: string }) => (
  <Paragraph
    leaf={{ attributes: {}, text: '' }}
    element={{ children: [{ text: '' }] }}
    text={{ text: '' }}
    attributes={{
      'data-slate-node': 'element',
      ref: null,
      'data-slate-leaf': true,
    }}
  >
    {args.text}
  </Paragraph>
);

Default.args = {
  text: 'Paragraph',
};

Default.storyName = 'Paragraph';
