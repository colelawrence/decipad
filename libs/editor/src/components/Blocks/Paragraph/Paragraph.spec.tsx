import { render } from 'test-utils';
import { Default } from './Paragraph.stories';

describe('Paragraph Block', () => {
  it('matches snapshot', () => {
    const { container } = render(<Default text="Paragraph" />);
    expect(container).toMatchSnapshot();
  });
});
