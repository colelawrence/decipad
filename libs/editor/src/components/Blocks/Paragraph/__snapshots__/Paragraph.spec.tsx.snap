// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph Block matches snapshot 1`] = `
.emotion-0 {
  line-height: 1.75;
  color: #3E3E42;
  padding: 8px 0;
  position: relative;
}

.emotion-0 ::selection {
  background-color: rgba(196, 202, 251, 0.5);
}

<div>
  <p
    class="emotion-0"
    data-slate-leaf="true"
    data-slate-node="element"
  >
    Paragraph
  </p>
</div>
`;
